src/CSnakes.Runtime/Python/Pack.cs

```diff
@@ -10,46 +15,233 @@ namespace CSnakes.Runtime.Python;
+
+    [DebuggerDisplay($"{{{nameof(DebuggerDisplay)}(),nq}}")]
+    struct RentalState
+    {
+        private bool returned;
+
+        /// <remarks>
+        /// This method should only be called from a state manager like <see
+        /// cref="RentedArray{T}"/>. It should not be called directly from user
+        /// code and thus is named "dangerous" to attract attention.
+        /// </remarks>
+        public void DangerousReturn() => returned = true;
+
+        public static implicit operator bool(RentalState b) => !b.returned;
+
+        private string DebuggerDisplay() => this ? "rented" : "returned";
+    }
+
+    [DebuggerDisplay($"{{{nameof(DebuggerDisplay)}(),nq}}")]
+    private ref struct RentedArray<T>
+    {
+        private readonly T[] array;
+        private ref RentalState rented;
+        private readonly ArrayPool<T> pool;
+        private readonly Span<T> span;
+
+        public RentedArray(ArrayPool<T> pool, int length, ref RentalState rental)
         {
-            var handles = items.Length < 18 // .NET tuples are max 17 items. This is a performance optimization.
-                ? stackalloc IntPtr[items.Length]
-                : new IntPtr[items.Length];
+            Debug.Assert(rental);
+            this.rented = ref rental;
+            this.pool = pool;
+            this.array = pool.Rent(length);
+            this.span = array.AsSpan(..length);
+        }
+
+        public int Length => Span.Length;
 
-            for (int i = 0; i < items.Length; i++)
+        private Span<T> Span
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
             {
-                SafeHandleMarshaller<PyObject>.ManagedToUnmanagedIn m = default;
-                m.FromManaged(items[i]);
-                marshallers.Add(m);
-                handles[i] = m.ToUnmanaged();
+                if (!this.rented)
+                {
+                    ThrowObjectDisposedException();
+                }
+
+                return this.span;
             }
-            return PyObject.Create(CPythonAPI.PackTuple(handles));
         }
-        finally
+
+        public void Dispose()
         {
-            foreach (var m in marshallers)
-            {
-                m.Free();
-            }
+            if (!this.rented)
+                return;
+
+            this.rented.DangerousReturn();
+            this.pool.Return(this.array);
         }
+
+        public Span<T>.Enumerator GetEnumerator() => Span.GetEnumerator();
+
+        public static implicit operator Span<T>(RentedArray<T> rented) => rented.Span;
+
+        private string DebuggerDisplay() =>
+            this.rented ? $"Length = {Length} (Capacity = {this.array.Length})" : "(returned)";
+
+        [DoesNotReturn]
+        private void ThrowObjectDisposedException() => throw new ObjectDisposedException(nameof(RentedArray<T>));
     }
 
-    internal static PyObject CreateList(Span<PyObject> items)
+    private static class ArrayPools
```

Comment:
The `ArrayPool` API is practically an "unsafe" API. The use case here doesn't seem unreasonable in practice, but is it needed? Do we have performance measurements that indicate the GC can't keep up with simply allocating a new array? Is memory pressure increasing in a manner that degrades performance?

Reply:
> In short, there's an approximate 10% cost to saving the memory for this profile of benchmarks.

I just saw the following. I don't think `ArrayPool` should be used for that little of a gain.

Reply:
@tonybaloney I think the purpose of this PR was achieved without going the extra distance of also optimising the array allocation through the various strategies (array pool, thread statics, etc.) that were attempted. I've reverted the PR to the last [approved](https://github.com/tonybaloney/CSnakes/pull/206#pullrequestreview-2314476562) state of 7f61cd1052da71cb8d3fc9ac220d9854108d0e1f with a1c1625a57684cc75b434cfcbd9b604ecebc0e57.

> > In short, there's an approximate 10% cost to saving the memory for this profile of benchmarks.
> 
> I just saw the following. I don't think `ArrayPool` should be used for that little of a gain.

I think we can debate whether throughput or memory footprint is the right call in a separate issue and once we have more feedback and data from usage in the wild/field. Meanwhile, hopefully this in a good state to merge?

Reply:
> I think we can debate whether throughput or memory footprint is the right call in a separate issue and once we have more feedback and data from usage in the wild/field.

Agree.